[Style] Stop special casing font properties in RenderStyle

weinig opened the pull request:
#### 97d0e362516b8739ab5739302541f7d7b150cbc8
<pre>
[Style] Stop special casing font properties in RenderStyle
<a href="https://bugs.webkit.org/show_bug.cgi?id=300159">https://bugs.webkit.org/show_bug.cgi?id=300159</a>

Reviewed by Darin Adler.

Requires all font properties to have getters/setters/initialValue
functions on RenderStyle, not just on FontCascadeDescription.

Adds strong style types for the `font-variant-*` types that
relate to multiple FontCascadeDescription properties to make
it possible for RenderStyle functions per-property.

Also removes the enum FontSmallCaps. It had no users.

* Source/WebCore/Headers.cmake:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/FontVariantBuilder.cpp: Removed.
* Source/WebCore/css/FontVariantBuilder.h: Removed.
* Source/WebCore/css/scripts/process-css-properties.py:
* Source/WebCore/platform/graphics/FontCascadeDescription.h:
* Source/WebCore/platform/graphics/FontDescription.cpp:
* Source/WebCore/platform/graphics/FontDescription.h:
* Source/WebCore/platform/text/TextFlags.cpp:
* Source/WebCore/platform/text/TextFlags.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleBuilderState.h:
* Source/WebCore/style/StyleBuilderStateInlines.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/values/fonts/StyleFontVariantAlternates.cpp: Added.
* Source/WebCore/style/values/fonts/StyleFontVariantAlternates.h: Added.
* Source/WebCore/style/values/fonts/StyleFontVariantEastAsian.cpp: Added.
* Source/WebCore/style/values/fonts/StyleFontVariantEastAsian.h: Added.
* Source/WebCore/style/values/fonts/StyleFontVariantLigatures.cpp: Added.
* Source/WebCore/style/values/fonts/StyleFontVariantLigatures.h: Added.
* Source/WebCore/style/values/fonts/StyleFontVariantNumeric.cpp: Added.
* Source/WebCore/style/values/fonts/StyleFontVariantNumeric.h: Added.
* Source/WebKit/Shared/TextFlags.serialization.in:
* Tools/Scripts/webkitpy/style/checkers/jsonchecker.py:

Canonical link: <a href="https://commits.webkit.org/301198@main">https://commits.webkit.org/301198@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ab2b3405ea4fa058d28b71c012e59c0dddb1c11e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/125189 "6 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/44856 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/35595 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/132040 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/77054 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/127066 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/45547 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/53417 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/95306 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/63246 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/128143 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/36362 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111950 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75846 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/124545 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/35260 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/30117 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/75518 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/106128 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/30343 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/134720 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51998 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39780 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103777 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/52433 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/108171 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/103547 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/26375 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48894 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/27184 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/49069 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51890 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/57669 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/51254 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/54610 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52947 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/style/values/fonts/StyleFontVariationSettings.cpp:
```diff
@@ -0,0 +1,152 @@
+
+auto CSSValueConversion<FontVariationSettings>::operator()(BuilderState& state, const CSSValue& value) -> FontVariationSettings
+{
+    if (RefPtr primitiveValue = dynamicDowncast<CSSPrimitiveValue>(value)) {
+        switch (primitiveValue->valueID()) {
+        case CSSValueNormal:
+            return CSS::Keyword::Normal { };
+        default:
+            if (CSSPropertyParserHelpers::isSystemFontShorthand(value.valueID()))
+                return CSS::Keyword::Normal { };
+            state.setCurrentPropertyInvalidAtComputedValueTime();
+            return CSS::Keyword::Normal { };
+        }
+    }
```
```suggestion
    if (RefPtr primitiveValue = dynamicDowncast<CSSPrimitiveValue>(value)) {
        switch (auto valueID = primitiveValue->valueID(); valueID) {
        case CSSValueNormal:
            return CSS::Keyword::Normal { };
        default:
            if (CSSPropertyParserHelpers::isSystemFontShorthand(valueID))
                return CSS::Keyword::Normal { };
            state.setCurrentPropertyInvalidAtComputedValueTime();
            return CSS::Keyword::Normal { };
        }
    }
```

darinadler commented on Source/WebCore/style/values/fonts/StyleFontFeatureSettings.cpp:
```diff
@@ -0,0 +1,112 @@
+                return CSS::Keyword::Normal { };
+            state.setCurrentPropertyInvalidAtComputedValueTime();
+            return CSS::Keyword::Normal { };
+        }
```
```suggestion
        switch (auto valueID = primitiveValue->valueID(); valueID) {
        case CSSValueNormal:
            return CSS::Keyword::Normal { };
        default:
            if (CSSPropertyParserHelpers::isSystemFontShorthand(valueID))
                return CSS::Keyword::Normal { };
            state.setCurrentPropertyInvalidAtComputedValueTime();
            return CSS::Keyword::Normal { };
        }
```

darinadler commented on Source/WebCore/platform/text/TextFlags.h:
```diff
@@ -375,26 +381,45 @@ struct FontVariantSettings {
+        : common(FontVariantLigatures::Normal)
+        , discretionary(FontVariantLigatures::Normal)
+        , historical(FontVariantLigatures::Normal)
+        , contextual(FontVariantLigatures::Normal)
```
Why not have these as default values on the data members themselves?

darinadler commented on Source/WebCore/platform/text/TextFlags.h:
```diff
@@ -375,26 +381,45 @@ struct FontVariantSettings {
+        , spacing(FontVariantNumericSpacing::Normal)
+        , fraction(FontVariantNumericFraction::Normal)
+        , ordinal(FontVariantNumericOrdinal::Normal)
+        , slashedZero(FontVariantNumericSlashedZero::Normal)
```
Why not have these as default values on the data members themselves?

darinadler commented on Source/WebCore/platform/text/TextFlags.h:
```diff
@@ -413,10 +438,19 @@ struct FontVariantNumericValues {
+    constexpr FontVariantEastAsianValues()
+        : variant(FontVariantEastAsianVariant::Normal)
+        , width(FontVariantEastAsianWidth::Normal)
+        , ruby(FontVariantEastAsianRuby::Normal)
```
Why not have these as default values on the data members themselves?

darinadler commented on LayoutTests/imported/w3c/web-platform-tests/css/css-fonts/animations/font-variation-settings-interpolation-expected.txt:
```diff
@@ -353,8 +353,8 @@ PASS Web Animations: property <font-variation-settings> from ['aaaa' 0, 'bbbb' 1
-FAIL CSS Transitions with transition: all: property <font-variation-settings> from ['aaaa' 30, 'bbbb' 20] to ['aaaa' 20, 'bbbb' 30] at (3.40282e+38) should be ['aaaa' -3.40282e+38, 'bbbb' 3.40282e+38] assert_array_equals: expected property 0 to be "\"aaaa\" -340282001837565597733306976381245063168" but got "\"aaaa\" -infinity" (expected array ["\"aaaa\" -340282001837565597733306976381245063168",  ...
-FAIL CSS Animations: property <font-variation-settings> from ['aaaa' 30, 'bbbb' 20] to ['aaaa' 20, 'bbbb' 30] at (3.40282e+38) should be ['aaaa' -3.40282e+38, 'bbbb' 3.40282e+38] assert_array_equals: expected property 0 to be "\"aaaa\" -340282001837565597733306976381245063168" but got "\"aaaa\" -infinity" (expected array ["\"aaaa\" -340282001837565597733306976381245063168", "\"bbbb\" 340282001837 ...
-FAIL Web Animations: property <font-variation-settings> from ['aaaa' 30, 'bbbb' 20] to ['aaaa' 20, 'bbbb' 30] at (3.40282e+38) should be ['aaaa' -3.40282e+38, 'bbbb' 3.40282e+38] assert_array_equals: expected property 0 to be "\"aaaa\" -340282001837565597733306976381245063168" but got "\"aaaa\" -infinity" (expected array ["\"aaaa\" -340282001837565597733306976381245063168", "\"bbbb\" 340282001837 ...
+FAIL CSS Transitions: property <font-variation-settings> from ['aaaa' 30, 'bbbb' 20] to ['aaaa' 20, 'bbbb' 30] at (3.40282e+38) should be ['aaaa' -3.40282e+38, 'bbbb' 3.40282e+38] assert_array_equals: expected property 0 to be "\"aaaa\" -340282001837565597733306976381245063168" but got "\"aaaa\" -340282346638528859811704183484516925440" (expected array ["\"aaaa\" -34028200183756559773330697638124 ...
+FAIL CSS Transitions with transition: all: property <font-variation-settings> from ['aaaa' 30, 'bbbb' 20] to ['aaaa' 20, 'bbbb' 30] at (3.40282e+38) should be ['aaaa' -3.40282e+38, 'bbbb' 3.40282e+38] assert_array_equals: expected property 0 to be "\"aaaa\" -340282001837565597733306976381245063168" but got "\"aaaa\" -340282346638528859811704183484516925440" (expected array ["\"aaaa\" -34028200183 ...
+FAIL CSS Animations: property <font-variation-settings> from ['aaaa' 30, 'bbbb' 20] to ['aaaa' 20, 'bbbb' 30] at (3.40282e+38) should be ['aaaa' -3.40282e+38, 'bbbb' 3.40282e+38] assert_array_equals: expected property 0 to be "\"aaaa\" -340282001837565597733306976381245063168" but got "\"aaaa\" -340282346638528859811704183484516925440" (expected array ["\"aaaa\" -340282001837565597733306976381245 ...
+FAIL Web Animations: property <font-variation-settings> from ['aaaa' 30, 'bbbb' 20] to ['aaaa' 20, 'bbbb' 30] at (3.40282e+38) should be ['aaaa' -3.40282e+38, 'bbbb' 3.40282e+38] assert_array_equals: expected property 0 to be "\"aaaa\" -340282001837565597733306976381245063168" but got "\"aaaa\" -340282346638528859811704183484516925440" (expected array ["\"aaaa\" -340282001837565597733306976381245 ...
```
Provocative failing results; I wonder what it would take to make this test pass.

darinadler commented on Source/WebCore/css/CSSFontFeatureValue.cpp:
```diff
@@ -44,8 +44,8 @@ String CSSFontFeatureValue::customCSSText(const CSS::SerializationContext& conte
     // Omit the value if it's `1` as `1` is implied by default.
-    if (m_value->resolveAsIntegerIfNotCalculated() != 1)
-        builder.append(' ', m_value->customCSSText(context));
+    if (RefPtr primitiveValue = dynamicDowncast<CSSPrimitiveValue>(m_value); !primitiveValue || primitiveValue->resolveAsIntegerIfNotCalculated() != 1)
```
I think it reads better reversed. Might be even better with a `isKnownOne` function.
```suggestion
    if (RefPtr primitiveValue = dynamicDowncast<CSSPrimitiveValue>(m_value); !(primitiveValue && primitiveValue->resolveAsIntegerIfNotCalculated() == 1))
```

darinadler commented on Source/WebCore/css/CSSFontFeatureValue.cpp:
```diff
@@ -44,8 +44,8 @@ String CSSFontFeatureValue::customCSSText(const CSS::SerializationContext& conte
     StringBuilder builder;
```
Kinda wish this used `makeString` instead of `StringBuilder`, but maybe in the longer term this will be in a function that takes a `StringBuilder&` anyway.

darinadler commented on Source/WebCore/platform/graphics/FontCustomPlatformData.h:
```diff
@@ -55,8 +55,8 @@ class FontDescription;
 
-template <typename T> class FontTaggedSettings;
-typedef FontTaggedSettings<int> FontFeatureSettings;
+template<typename T> class FontTaggedSettings;
```
```suggestion
template<typename> class FontTaggedSettings;
```

darinadler commented on Source/WebCore/platform/graphics/FontDescription.h:
```diff
@@ -86,6 +86,33 @@ class FontDescription {
+            variantHistoricalLigatures(),
+            variantContextualAlternates(),
+        };
+    }
```
I would put these vertically formatted function bodies outside the class definition.

darinadler commented on Source/WebCore/platform/graphics/FontDescription.h:
```diff
@@ -149,6 +176,27 @@ class FontDescription {
+        setVariantDiscretionaryLigatures(values.discretionary);
+        setVariantHistoricalLigatures(values.historical);
+        setVariantContextualAlternates(values.contextual);
+    }
```
I would put these vertically formatted function bodies outside the class definition.

darinadler commented on Source/WebCore/platform/graphics/FontTaggedSettings.h:
```diff
@@ -68,15 +68,15 @@ struct FourCharacterTagHashTraits : HashTraits<FontTag> {
 
-    friend bool operator==(const FontTaggedSetting&, const FontTaggedSetting&) = default;
+    bool operator==(const FontTaggedSetting&) const = default;
     bool operator<(const FontTaggedSetting<T>& other) const;
```
```suggestion
    bool operator<(const FontTaggedSetting&) const;
```
Also, maybe this should be a `<=>` instead of `<` operator.

darinadler commented on Source/WebCore/rendering/style/RenderStyle.h:
```diff
@@ -728,12 +739,28 @@ class RenderStyle final : public CanMakeCheckedPtr<RenderStyle> {
 
     inline FontOpticalSizing fontOpticalSizing() const;
-    inline FontVariationSettings fontVariationSettings() const;
+    inline Style::FontFeatureSettings fontFeatureSettings() const;
```
In such a massive class, I’d prefer alphabetical order rather than a “logical” order. Something to consider later, I suppose.

weinig replied:
Once I get all the properties in (very soon), I will sort all them.

darinadler reviewed (approved):
Not just this pull request: Not sure how to feel about large functions in header files.
1. Given we have link-time optimization, are we slowing down compilation without sufficient performance benefit?
2. Are we losing the benefit of the class definition as clarifying the interface because we are interspersing it with so much of the implementation?
3. One possible benefit is if we don’t need .cpp files at all, nice not to have twice as many source files I suppose.

weinig committed:
[Style] Stop special casing font properties in RenderStyle
https://bugs.webkit.org/show_bug.cgi?id=300159

Reviewed by Darin Adler.

Requires all font properties to have getters/setters/initialValue
functions on RenderStyle, not just on FontCascadeDescription.

Adds strong style types for the `font-variant-*` types that
relate to multiple FontCascadeDescription properties to make
it possible for RenderStyle functions per-property.

Also removes the enum FontSmallCaps. It had no users.

* Source/WebCore/Headers.cmake:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/FontVariantBuilder.cpp: Removed.
* Source/WebCore/css/FontVariantBuilder.h: Removed.
* Source/WebCore/css/scripts/process-css-properties.py:
* Source/WebCore/platform/graphics/FontCascadeDescription.h:
* Source/WebCore/platform/graphics/FontDescription.cpp:
* Source/WebCore/platform/graphics/FontDescription.h:
* Source/WebCore/platform/text/TextFlags.cpp:
* Source/WebCore/platform/text/TextFlags.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleBuilderState.h:
* Source/WebCore/style/StyleBuilderStateInlines.h:
* Source/WebCore/style/StyleExtractorCustom.h:
* Source/WebCore/style/StyleInterpolationWrappers.h:
* Source/WebCore/style/values/fonts/StyleFontVariantAlternates.cpp: Added.
* Source/WebCore/style/values/fonts/StyleFontVariantAlternates.h: Added.
* Source/WebCore/style/values/fonts/StyleFontVariantEastAsian.cpp: Added.
* Source/WebCore/style/values/fonts/StyleFontVariantEastAsian.h: Added.
* Source/WebCore/style/values/fonts/StyleFontVariantLigatures.cpp: Added.
* Source/WebCore/style/values/fonts/StyleFontVariantLigatures.h: Added.
* Source/WebCore/style/values/fonts/StyleFontVariantNumeric.cpp: Added.
* Source/WebCore/style/values/fonts/StyleFontVariantNumeric.h: Added.
* Source/WebKit/Shared/TextFlags.serialization.in:
* Tools/Scripts/webkitpy/style/checkers/jsonchecker.py:

Canonical link: https://commits.webkit.org/301198@main
